### ISSUE TITLE
Workflow metrics with workflow using DAG do not show in workflow controller /metrics

### ISSUE THREAD

**Issue by pacospace** [argoproj/argo-workflows#3941]
## Summary 

What happened/what you expected to happen?

Recently introduced metrics in Argo workflows to monitor duration and status for a specific task. But although workflow is successful, no metrics are shown in the workflow controller

Main Workflow: https://github.com/thoth-station/thoth-application/blob/master/solver/base/openshift-templates/solve_and_sync-template.yaml

WorkflowTemplate (task from above workflow): https://github.com/thoth-station/thoth-application/blob/master/solver/base/argo-workflows/solver-template.yaml

## Diagnostics

What version of Argo Workflows are you running?

v2.9.5

Workflow run successfully

Part of the workflow linked above
```yaml
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: solver-any
spec:
  templates:
    - name: solve
      metrics:
        prometheus:
          - name: exec_duration_gauge
            help: "Duration task gauge by name when succeded."
            labels:
              - key: name
                value: "{{workflow.name}}"
              - key: workflow_type
                value: "solver"
            when: "{{status}} == Succeeded"
            gauge:
              value: "{{duration}}"
          - name: status_counter
            labels:
              - key: name
                value: "solver"
              - key: status
                value: "{{status}}"
            help: "Count of workflow by status"
            counter:
              value: "1"
```

```
No logs relative to metrics
```

---
**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by simster7** [comment]
Could you please add a minimal Workflow that has this problem that I could easily use to reproduce?

**Comment by pacospace** [comment]
> Could you please add a minimal Workflow that has this problem that I could easily use to reproduce?

I tested two cases, I think the issue is when DAG are used (at least with this test).

I have two simple versions:

This one works:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: hello-world-
spec:
  serviceAccountName: argo
  entrypoint: whalesay
  templates:
    - name: whalesay
      metrics:
        prometheus:
          - name: execution_counter
            help: "How many times a step has executed"
            labels:
              - key: name
                value: flakey
            counter:
              value: "1"
          - name: failure_counter
            help: "How many times a step has failed"
            labels:
              - key: name
                value: flakey
            when: "{{status}} == Error"
            counter:
              value: "1"
      container:
        image: docker/whalesay:latest
        command: [cowsay]
```

Using DAG does not work instead:

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: hello-world-
spec:
  serviceAccountName: argo
  entrypoint: whalesay
  templates:
    - name: whalesay
      dag:
        tasks:
          - name: whalesay
            template: whalesay-dag
    - name: whalesay-dag
      metrics:
        prometheus:
          - name: execution_counter
            help: "How many times a step has executed"
            labels:
              - key: name
                value: test-dag
            counter:
              value: "1"
          - name: failure_counter
            help: "How many times a step has failed"
            labels:
              - key: name
                value: test-dag
            when: "{{status}} == Error"
            counter:
              value: "1"
      container:
        image: docker/whalesay:latest
        command: [cowsay]
```

**Comment by simster7** [comment]
Just confirmed that this is fixed by #3886. The fix should be on the next patch release.

**Comment by pacospace** [comment]
> Just confirmed that this is fixed by #3886. The fix should be on the next patch release.

Thanks a lot @simster7 ! Looking forward to that
